src/dotty/tools/dotc/transform/LambdaLift.scala

```diff
@@ -195,17 +245,16 @@ class LambdaLift extends MiniPhase with IdentityDenotTransformer { thisTransform
         }
         tree match {
           case tree: Ident =>
-            if (sym.maybeOwner.isTerm) {
+            if (isLocal(sym)) {
               if (sym is Label)
                 assert(enclosure == sym.enclosure,
                   i"attempt to refer to label $sym from nested $enclosure")
               else if (sym is Method) markCalled(sym, enclosure)
               else if (sym.isTerm) markFree(sym, enclosure)
-            } else if (sym.maybeOwner.isClass)
-              narrowTo(sym.owner.asClass)
+            }
+            if (sym.maybeOwner.isClass) narrowTo(sym.owner.asClass)
```

Comment:
If the owner of a symbol is a class, then previously that meant the symbol has to be non-local. Now, it could be local, but then it's either owned by a trait or is a constructor. In both cases we
do not want to move the symbol out of the enclosing class.